### PR TITLE
std: Stabilize the `fs` module

### LINKED ISSUE #22879: Stop statting paths before opening them
https://github.com/rust-lang/rust/blob/master/src/libstd/fs/mod.rs#L275-L280

It's ruining my ability to write a sandbox. The `stat` syscall might be blacklisted while `open` is not. That should not prevent you from opening files!

Nominating because I believe this must be fixed before 1.0.

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit performs a stabilization pass over the `std::fs` module now that
it's had some time to bake. The change was largely just adding `#[stable]` tags,
but there are a few APIs that remain `#[unstable]`.

The following apis are now marked `#[stable]`:
- `std::fs` (the name)
- `File`
- `Metadata`
- `ReadDir`
- `DirEntry`
- `OpenOptions`
- `Permissions`
- `File::{open, create}`
- `File::{sync_all, sync_data}`
- `File::set_len`
- `File::metadata`
- Trait implementations for `File` and `&File`
- `OpenOptions::new`
- `OpenOptions::{read, write, append, truncate, create}`
- `OpenOptions::open` - this function was modified, however, to not attempt to
  reject cross-platform openings of directories. This means that some platforms
  will succeed in opening a directory and others will fail.
- `Metadata::{is_dir, is_file, len, permissions}`
- `Permissions::{readonly, set_readonly}`
- `Iterator for ReadDir`
- `DirEntry::path`
- `remove_file` - like with `OpenOptions::open`, the extra windows code to
  remove a readonly file has been removed. This means that removing a readonly
  file will succeed on some platforms but fail on others.
- `metadata`
- `rename`
- `copy`
- `hard_link`
- `soft_link`
- `read_link`
- `create_dir`
- `create_dir_all`
- `remove_dir`
- `remove_dir_all`
- `read_dir`

The following apis remain `#[unstable]`.
- `WalkDir` and `walk` - there are many methods by which a directory walk can be
  constructed, and it's unclear whether the current semantics are the right
  ones. For example symlinks are not handled super well currently. This is now
  behind a new `fs_walk` feature.
- `File::path` - this is an extra abstraction which the standard library
  provides on top of what the system offers and it's unclear whether we should
  be doing so. This is now behind a new `file_path` feature.
- `Metadata::{accessed, modified}` - we do not currently have a good
  abstraction for a moment in time which is what these APIs should likely be
  returning, so these remain `#[unstable]` for now. These are now behind a new
  `fs_time` feature
- `set_file_times` - like with `Metadata::accessed`, we do not currently have
  the appropriate abstraction for the arguments here so this API remains
  unstable behind the `fs_time` feature gate.
- `PathExt` - the precise set of methods on this trait may change over time and
  some methods may be removed. This API remains unstable behind the `path_ext`
  feature gate.
- `set_permissions` - we may wish to expose a more granular ability to set the
  permissions on a file instead of just a blanket "set all permissions" method.
  This function remains behind the `fs` feature.

The following apis are now `#[deprecated]`
- The `TempDir` type is now entirely deprecated and is [located on
  crates.io](https://crates.io/crates/tempdir) as the `tempdir` crate with [its source](https://github.com/rust-lang/tempdir) at
  rust-lang/tempdir.

The stability of some of these APIs has been questioned over the past few weeks
in using these APIs, and it is intentional that the majority of APIs here are
marked `#[stable]`. The `std::fs` module has a lot of room to grow and the
material is [being tracked in a RFC issue](rust-lang/rfcs#939).

Closes #22879 

[breaking-change]

### CODE REVIEW

**Comment by aturon** [comment]
@alexcrichton Would you mind putting the remarks about bits you're keeping as unstable in the stability reason itself? That helps us keep track of what needs to be resolved before stabilization.

File: src/libstd/fs/mod.rs [hunk 1]
```diff
@@ -77,9 +81,11 @@ pub struct ReadDir(fs_imp::ReadDir);
 /// An instance of `DirEntry` represents an entry inside of a directory on the
 /// filesystem. Each entry can be inspected via methods to learn about the full
 /// path or possibly other metadata through per-platform extension traits.
+#[stable(feature = "fs", since = "1.0.0")]
 pub struct DirEntry(fs_imp::DirEntry);
 
 /// An iterator that recursively walks over the contents of a directory.
+#[unstable(feature = "fs_walk")]
```
Comment (aturon) [inline src/libstd/fs/mod.rs]: Can you add in the reason that you have in the commit message here?

File: src/libstd/fs/mod.rs [hunk 2]
```diff
@@ -122,11 +131,13 @@ impl File {
     /// and will truncate it if it does.
     ///
     /// See the `OpenOptions::open` function for more details.
+    #[stable(feature = "fs", since = "1.0.0")]
     pub fn create<P: AsPath + ?Sized>(path: &P) -> io::Result<File> {
         OpenOptions::new().write(true).create(true).truncate(true).open(path)
     }
 
     /// Returns the original path that was used to open this file.
+    #[unstable(feature = "file_path")]
```
Comment (aturon) [inline src/libstd/fs/mod.rs]: Same here.

File: src/libstd/fs/mod.rs [hunk 3]
```diff
@@ -159,11 +172,13 @@ impl File {
     /// be shrunk. If it is greater than the current file's size, then the file
     /// will be extended to `size` and have all of the intermediate data filled
     /// in with 0s.
+    #[stable(feature = "fs", since = "1.0.0")]
     pub fn set_len(&self, size: u64) -> io::Result<()> {
         self.inner.truncate(size)
     }
 
     /// Queries information about the underlying file.
```
Comment (aturon) [inline src/libstd/fs/mod.rs]: nit: the doc comment should probably say: "Queries metadata about..."

File: src/libstd/fs/mod.rs [hunk 4]
```diff
@@ -264,37 +291,33 @@ impl OpenOptions {
...
+    #[stable(feature = "fs", since = "1.0.0")]
     pub fn open<P: AsPath + ?Sized>(&self, path: &P) -> io::Result<File> {
         let path = path.as_path();
         let inner = try!(fs_imp::File::open(path, &self.0));
-
-        // On *BSD systems, we can open a directory as a file and read from
-        // it: fd=open("/tmp", O_RDONLY); read(fd, buf, N); due to an old
-        // tradition before the introduction of opendir(3).  We explicitly
-        // reject it because there are few use cases.
-        if cfg!(not(any(target_os = "linux", target_os = "android"))) &&
```
Comment (aturon) [inline src/libstd/fs/mod.rs]: cc @pcwalton, the `stat` is being removed.

File: src/libstd/fs/mod.rs [hunk 5]
```diff
@@ -649,11 +668,13 @@ pub fn read_dir<P: AsPath + ?Sized>(path: &P) -> io::Result<ReadDir> {
 ///
 /// The iterator will yield instances of `io::Result<DirEntry>`. New errors may
 /// be encountered after an iterator is initially constructed.
+#[unstable(feature = "fs_walk")]
```
Comment (aturon) [inline src/libstd/fs/mod.rs]: Please add reason.

File: src/libstd/fs/mod.rs [hunk 6]
```diff
@@ -727,6 +754,7 @@ impl PathExt for Path {
 /// The file at the path specified will have its last access time set to
 /// `atime` and its modification time set to `mtime`. The times specified should
 /// be in milliseconds.
+#[unstable(feature = "fs_time")]
```
Comment (aturon) [inline src/libstd/fs/mod.rs]: add reason

File: src/libstd/fs/mod.rs [hunk 7]
```diff
@@ -752,6 +780,7 @@ pub fn set_file_times<P: AsPath + ?Sized>(path: &P, accessed: u64,
 /// This function will return an error if the provided `path` doesn't exist, if
 /// the process lacks permissions to change the attributes of the file, or if
 /// some other I/O error is encountered.
+#[stable(feature = "fs", since = "1.0.0")]
 pub fn set_permissions<P: AsPath + ?Sized>(path: &P, perm: Permissions)
```
Comment (aturon) [inline src/libstd/fs/mod.rs]: Hm, I wonder whether we should consider a builder-style API here instead?

It's a little unclear how extensive permissions will get, and whether in all cases you'll want to clobber the entire current permissions with an instance of `Permissions`. A builder could make it easier to say exactly what you want to override. Thoughts?

**Comment by aturon** [comment]
OK, I've left a few nits, but otherwise r=me.

**Comment by blaenk** [comment]
good thing I saw this, now I know that tempdir is a separate crate. sounds good to me :D

**Pull request merged**